### PR TITLE
restarting a kubelet should never affect the running workload

### PR DESCRIPTION
**Pull request by yangyefei90**
#### What type of PR is this?

Add one of the following kinds:
/kind bug

#### Which issue(s) this PR fixes:
Usage:
Fixes # https://github.com/kubernetes/kubernetes/issues/123980

### CODE REVIEW

**Comment by yangyefei90** [comment]
The admission check for kubelet should be a one-time thing. For pods that have been admitted, no matter how the node configuration changes, it should not affect the pods that have been admitted. There are two main points to check here. One is whether there is an IP. If the IP has been assigned, it proves that the pod already has a sandbox and must have been admitted. The other is whether the pod is in the pending state. The pod in the non-pending state must have been processed by kubelet, because from pending to non- The conversion of Pending state is done by kubelet.

**Comment by yangyefei90** [comment]
@bart0sh please help review，thanks

**Comment by gjkim42** [comment]
Could you explain more about the real buggy cases we should address with this PR?

> The admission check for kubelet should be a one-time thing. For pods that have been admitted, no matter how the node configuration changes, it should not affect the pods that have been admitted. There are two main points to check here. One is whether there is an IP. If the IP has been assigned, it proves that the pod already has a sandbox and must have been admitted. The other is whether the pod is in the pending state. The pod in the non-pending state must have been processed by kubelet, because from pending to non- The conversion of Pending state is done by kubelet.

I couldn't fully agree with this.
I would say restarting a kubelet should not affect the most running workloads however there could be a workload that we want to evict with updating the kubelet configuration.

**Comment by ffromani** [comment]
> Could you explain more about the real buggy cases we should address with this PR?
> 
> > The admission check for kubelet should be a one-time thing. For pods that have been admitted, no matter how the node configuration changes, it should not affect the pods that have been admitted. There are two main points to check here. One is whether there is an IP. If the IP has been assigned, it proves that the pod already has a sandbox and must have been admitted. The other is whether the pod is in the pending state. The pod in the non-pending state must have been processed by kubelet, because from pending to non- The conversion of Pending state is done by kubelet.
> 
> I couldn't fully agree with this. I would say restarting a kubelet should not affect the MOST running workloads however there could be a workload that we want to evict by updating the kubelet configuration.

I think that in general a kubelet restart should not affect the resource allocation of a running workload (e.g. devices, exclusive cpus)

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -2577,21 +2580,25 @@ func (kl *Kubelet) HandlePodAdditions(pods []*v1.Pod) {
...
-				// Check if we can admit the pod; if not, reject it.
-				if ok, reason, message := kl.canAdmitPod(activePods, podCopy); !ok {
-					kl.rejectPod(pod, reason, message)
-					continue
+				if needToAdmit(pod) {
+					// Check if we can admit the pod; if not, reject it.
+					if ok, reason, message := kl.canAdmitPod(activePods, podCopy); !ok {
+						kl.rejectPod(pod, reason, message)
+						continue
+					}
```
Comment (gjkim42) [inline pkg/kubelet/kubelet.go:2588]: Regardless of whether the kubelet restart should not affect all cases or not, this change is not desirable.

`canAdmitPod` also includes the eviction process of the kubelet, we cannot disable it for all running pods. Kubelet should be able to evict running pods for various reasons (ephemeral storage pressure, ...).

You first need to define concrete buggy cases and address them one by one.

**Review by gjkim42 — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (ffromani) [reply pkg/kubelet/kubelet.go:2588]: very good point about eviction and the eviction manager

Comment (AllenXu93) [reply pkg/kubelet/kubelet.go:2590]: Evict running pods won't get in this func,  HandlePodAdditions only handle pods for new created or kubelet restart;
This change will only affect for running pods when kubelet restart.

**Pull request closed without merging**